File: LongestHappyPrefix.py
class Solution:
    def piTable(self, pat):
        n = len(pat)
        pi = [0] * n

        i, length = 1, 0

        while i < n:
            if pat[i] == pat[length]:
                length += 1
                pi[i] = length
                i += 1
            else:
                if length != 0:
                    length = pi[length - 1]
                else:
                    pi[i] = 0
                    i += 1

        return pi

    def longestPrefix(self, s):
        pi = self.piTable(s)

        maxLen = pi[len(s) - 1]

        res = ""

        for char in s:
            if maxLen == 0:
                break
            res += char
            maxLen -= 1

        return res

# Example usage:
solution = Solution()
s = "ababab"
longest_prefix = solution.longestPrefix(s)
print("Longest Prefix:", longest_prefix)

File: LongestHappyPrefix.cpp
class Solution {
public:
    vector<int>piTable(string pat){
          vector<int>pi;
          int n = pat.length();
          pi.resize(n);
          
          int i = 1, len = 0;

          while(i < n){
                if(pat[i] == pat[len]){
                     len++;
                     pi[i] = len;
                     i++;
                }else{
                      
                       if(len != 0){
                            len = pi[len-1];
                       }else{
                            
                            pi[i] = 0;
                            i++;
                       }
                }
          }

          return pi;
    }
    string longestPrefix(string s) {
         vector<int>pi = piTable(s);

         int maxLen = pi[s.length()-1];

         string res = "";

         for(auto it: s){
               if(maxLen == 0)break;
               res += it;
               maxLen--;
         }

         return res;
    }
};
